zebra-state/src/service/finalized_state/disk_format/upgrade.rs

```diff
@@ -261,46 +263,165 @@ impl DbFormatChange {
+                while let Ok(height) = unique_sapling_tree_height_rx.recv() {
+                    let delete_from = (prev_height + 1).expect("should be valid height");
+                    let num_entries = height.0.checked_sub(delete_from.0);
+
+                    if num_entries.map_or(false, |n| n >= 1) {
+                        let mut batch: DiskWriteBatch = DiskWriteBatch::new();
+                        batch.delete_range_sapling_tree(&db, &delete_from, &height);
+                        db.write_batch(batch).expect(
+                            "Deleting Sapling note commitment trees should always succeed.",
+                        );
+                    }
+
+                    prev_height = height;
+                }
+            });
+
+            // Create an unbounded channel for reading note commitment trees
+            let (orchard_tree_tx, orchard_tree_rx) = mpsc::channel();
+
+            // Set up task for reading orchard note commitment trees
+            let db = upgrade_db.clone();
+            let orchard_read_task = std::thread::spawn(move || {
+                for (height, tree) in
+                    db.orchard_tree_by_height_range(orchard_height..initial_tip_height)
+                {
+                    if let Err(error) = orchard_tree_tx.send((height, tree)) {
+                        warn!(?error, "unexpected send error")
+                    }
+                }
+            });
+
+            // Create an unbounded channel for duplicate orchard note commitment tree heights
+            let (unique_orchard_tree_height_tx, unique_orchard_tree_height_rx) = mpsc::channel();
+
+            // Set up task for reading orchard note commitment trees
+            let orchard_compare_task = std::thread::spawn(move || {
+                while let Ok((height, tree)) = orchard_tree_rx.recv() {
+                    let tree = Some(tree);
+                    if prev_orchard_tree != tree {
+                        if let Err(error) = unique_orchard_tree_height_tx.send(height) {
+                            warn!(?error, "unexpected send error")
+                        }
+
+                        prev_orchard_tree = tree;
+                    }
+                }
+            });
+
+            // Set up task for deleting orchard note commitment trees
+            let db = upgrade_db;
+            let orchard_delete_task = std::thread::spawn(move || {
+                let mut prev_height = orchard_height;
+                while let Ok(height) = unique_orchard_tree_height_rx.recv() {
+                    let delete_from = (prev_height + 1).expect("should be valid height");
+                    let num_entries = height.0.checked_sub(delete_from.0);
+
+                    if num_entries.map_or(false, |n| n >= 1) {
+                        let mut batch: DiskWriteBatch = DiskWriteBatch::new();
+                        batch.delete_range_orchard_tree(&db, &delete_from, &height);
```

Comment:
Nitpick: it may be slightly faster to use `zs_delete` when the range would only delete a single key.

Reply:
Wait but with that said, great cleanup - please bring it back without the duplicate `batch` definitions and `write_batch` calls.

Reply:
Done in [235e43b](https://github.com/ZcashFoundation/zebra/pull/7312/commits/235e43bf8f98a7442cc885147a42c03102d0a07b).